tools/total-contributions

```diff
@@ -0,0 +1,158 @@
+#!/usr/bin/env python3
+import argparse
+import os
+import pathlib
+import subprocess
+import sys
+from datetime import datetime
+from typing import Dict, List
+
+import requests
+
+
+def add_log(dict: Dict[str, int], input: List[str]) -> None:
+    for dataset in input:
+        committer_name = dataset.split("\t")[1]
+        commit_count = int(dataset.split("\t")[0])
+        if committer_name in dict:
+            dict[committer_name] += commit_count
+        else:
+            dict[committer_name] = commit_count
+
+
+def retrieve_log(repo: str, lower_version: str, upper_version: str) -> List[str]:
+    return subprocess.check_output(
+        ["git", "shortlog", "-s", lower_version + ".." + upper_version],
+        universal_newlines=True,
+        cwd=find_path(repo),
+    ).splitlines()
+
+
+def find_version(time: str, repo: str) -> str:
+    """
+    Versioning in zulip main repo and other sub repos are different. When user inputs a version number, it corresponds to the version of the main zulip repo. Using the timestamp of the main repo's release version, we find the version of the other sub repos at that timestamp.
+
+    Args:
+        time (str): timestamp corresponding to main zulip repo's release version
+        repo (str): name of the repo whose release version has to be found
+
+    Returns:
+        version (str): release version of that particular 'repo'
+    """
+    versions_list = subprocess.check_output(
+        ["git", "tag", "-l"], universal_newlines=True, cwd=find_path(repo)
+    ).splitlines()
+    for version in versions_list:
+        version_time = subprocess.check_output(
+            ["git", "log", "-1", "--format=%ai", version],
+            universal_newlines=True,
+            cwd=find_path(repo),
+        ).split()[0]
+        if datetime.strptime(version_time, "%Y-%m-%d") >= datetime.strptime(time, "%Y-%m-%d"):
+            return version
+    return version
+
+
+def find_path(repo: str) -> str:
+    return os.path.dirname((pathlib.Path().resolve()).parents[0]) + "/" + repo
+
+
+# get latest release version
+def get_latest_version() -> str:
+    try:
+        response = requests.get("https://api.github.com/repos/zulip/zulip/releases/latest")
+        return response.json()["tag_name"]
+    except requests.exceptions.RequestException:
```

Comment:
And the user needs to know what the exception is.

Reply:
Instead of exit shall I stdout a warning about the type of exception. Then mention that due to the exception, the release version number from the local repo is being used?

Reply:
Also, if we exit early with ```exit(1)```  the case in lines 66 - 70 will never be executed. Can you please let me know if this is the expected behavior?

Reply:
I suppose, as long as there is a clear warning message, it's fine to show the local version.

Reply:
By the way, what is the reason for using api.github.com instead of just using the local repo tags? I think you'd have to make sure that the repos are up-to-date anyway before running this script?

Reply:
@rht When a user doesn't know or wants to get the data till the latest version. He doesn't have to enter the full version tag, instead, he could just run the script as ```total-contributions 2.0```  which will fetch data from version 2.0 till the latest version. This latest version tag is received from api.github.com. Now as you had mentioned, if there are any connectivity issues we are using the local repo tag as the latest version.

Reply:
I meant, fetching from api.github.com is redundant if the repos have already been made sure to be up-to-date.